pkg/controller/datavolume/controller-base.go

```diff
@@ -323,18 +346,123 @@ func (r ReconcilerBase) syncUpdate(log logr.Logger, syncRes *dataVolumeSyncResul
 	return nil
 }
 
-func (r ReconcilerBase) handlePrePopulation(dv *cdiv1.DataVolume, pvc *corev1.PersistentVolumeClaim) {
+func (r *ReconcilerBase) handleStaticVolume(syncRes *dataVolumeSyncResult, log logr.Logger) error {
+	if _, ok := syncRes.dvMutated.Annotations[cc.AnnCheckStaticVolume]; !ok {
+		return nil
+	}
+
+	if syncRes.pvc == nil {
+		volumes, err := r.getAvailableVolumesForDV(syncRes, log)
+		if err != nil {
+			return err
+		}
+
+		if len(volumes) == 0 {
+			log.Info("No PVs for DV")
+			return nil
+		}
+
+		if err := r.handlePvcCreation(log, syncRes, func(_ *cdiv1.DataVolume, pvc *corev1.PersistentVolumeClaim) error {
+			bs, err := json.Marshal(volumes)
+			if err != nil {
+				return err
+			}
+			cc.AddAnnotation(pvc, cc.AnnPersistentVolumeList, string(bs))
+			return nil
+		}); err != nil {
+			return err
+		}
+
+		// set result to make sure callers don't do anything else in sync
+		syncRes.result = &reconcile.Result{}
+		return nil
+	}
+
+	volumeAnno, ok := syncRes.pvc.Annotations[cc.AnnPersistentVolumeList]
+	if !ok {
+		// etiher did not create the PVC here OR bind to expected PV succeeded
+		return nil
+	}
+
+	if syncRes.pvc.Spec.VolumeName == "" {
+		// set result to make sure callers don't do anything else in sync
+		syncRes.result = &reconcile.Result{}
+		return nil
+	}
+
+	var volumes []string
+	if err := json.Unmarshal([]byte(volumeAnno), &volumes); err != nil {
+		return err
+	}
+
+	for _, v := range volumes {
+		if v == syncRes.pvc.Spec.VolumeName {
+			pvcCpy := syncRes.pvc.DeepCopy()
+			// handle as "populatedFor" going foreward
+			cc.AddAnnotation(pvcCpy, cc.AnnPopulatedFor, syncRes.dvMutated.Name)
+			delete(pvcCpy.Annotations, cc.AnnPersistentVolumeList)
+			if err := r.updatePVC(pvcCpy); err != nil {
+				return err
+			}
+			syncRes.pvc = pvcCpy
+			return nil
+		}
+	}
+
+	// delete the pvc and hope for better luck...
+	pvcCpy := syncRes.pvc.DeepCopy()
+	if err := r.client.Delete(context.TODO(), pvcCpy, &client.DeleteOptions{}); err != nil {
+		return err
+	}
+
+	syncRes.pvc = pvcCpy
+
+	return fmt.Errorf("DataVolume bound to unexpected PV %s", syncRes.pvc.Spec.VolumeName)
```

Comment:
If I understand correctly by reading the PR description, shouldn't we return to step 1 in this case? isn't this an irrecoverable state?

Reply:
We do.  Next reconcile there will be no pvc and we do the same steps over again

Reply:
Nice! I thought that we only returned reconcile errors in irrecoverable states.

Reply:
Nah, whenever reconcile returns an error it is retried with exponential backoff until reconcile returns no error